Add "Delay" effect

mofeing opened the pull request:
Adding a simple Delay effect.
The buffer allocation responsibility is delegated to the user of the lib (to avoid heap allocation and keep compatibility with no_std and embedded).

mofeing committed:
Init "effects" module

mofeing committed:
Delete unused trait in "effects" base module

mofeing committed:
Add "Delay" effect

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,77 @@
+use crate::float::Float;
+use std::collections::VecDeque;
```
We need to use a collection that doesn't use the heap to keep compatibility with microcontrollers.

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,77 @@
+    self.buffer.push_back(input);
+  }
+
+  pub fn get(&self, index: F) -> F {
```
the unit for index is `seconds` right ? can we change the name to avoid confusion with an actual index in the buffer ?

mofeing replied:
We could rename it with `time` for example

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,77 @@
+    }
+  }
+
+  pub fn set_delay(&mut self, delay: F) {
```
how do we make sure that it never has a value of 0 ?

mofeing replied:
If we want to check it here, we have 3 options:
- Return a `Result<(), E>`
- `panic!`
- Ignore and don't change

But I guess that the best thing is to put limits in the component wrapping this code (e.g. min and max values for the parameters).

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  pub fn process(&mut self, input: F) -> F {
+    let sample = self.delayline.get(self.delay);
```
the multiplication to convert `seconds` to a sample `index` is done for every process call.
I propose we store the conversion between `seconds` to `samples` when the `delay` parameter changes, then all the internal calculations are done with indices.

mofeing replied:
I did this way at first but wanted to first implement it like this for clarity.

mofeing committed:
Refactor DelayLine

- DelayLine buffer is allocated externally
- Delay is now the responsible for the time - discrete samples conversion

mofeing committed:
Test DelayLine

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,127 @@
+    let sample = self.delayline.get(self.delay_samples);
+    self.delayline.update(input + sample * self.feedback);
+
+    sample * self.mix + input * (F::val(1.0) - self.mix)
```
```suggestion
    sample * self.mix + input * (F::one() - self.mix)
```

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,127 @@
+  pub fn new(sample_rate: F, buffer: &'a mut [F]) -> Self {
+    Self {
+      delay_samples: 1, // delay in samples NOTE it must never be 0
+      delay_seconds: F::val(1) / sample_rate,
```
```suggestion
      delay_seconds: sample_rate.recip(),
```

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,127 @@
+  }
+}
+
+pub struct Delay<'a, F: Float> {
+  delay_samples: usize,
+  delay_seconds: F,
+  feedback: F,
+  mix: F,
+  delayline: DelayLine<'a, F>,
+  sample_rate: F,
+}
```
```suggestion
/// Simple delay effect with 3 parameters: the delay amount, the amount of feedback and dry/run mix.
pub struct Delay<'a, F: Float> {
  /// The amount of delay for the output signal. Values from `1.0 / sample_rate` to `buffer.len() / sample_rate`.
  delay_seconds: F,
  /// The amount of feedback for the output signal into the delay line again. Values from 0.0 to 1.0
  feedback: F,
  /// The dry/wet proportion. Values from 0.0 (dry) to 1.0 (wet)
  mix: F,
  
  sample_rate: F,
  delayline: DelayLine<'a, F>,
  delay_samples: usize,
}
```

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,127 @@
+impl<'a, F: Float> Delay<'a, F> {
+  pub fn new(sample_rate: F, buffer: &'a mut [F]) -> Self {
+    Self {
+      delay_samples: 1, // delay in samples NOTE it must never be 0
```
```suggestion
      delay_samples: 1,
```

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,127 @@
+    Self {
+      delay_samples: 1, // delay in samples NOTE it must never be 0
+      delay_seconds: F::val(1) / sample_rate,
+      feedback: F::zero(), // between 0 and 1
+      mix: F::zero(),      // between 0 and 1
```
```suggestion
      feedback: F::zero(),
      mix: F::zero(),
```

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,127 @@
+  #[test]
+  fn delayline_get() {
+    let mut buffer = [4., 3., 2., 1.];
+    // let delayline = DelayLine::new(1.0, &mut buffer);
```
```suggestion
```

chris-zen committed:
Apply suggestions from code review

- comments
- F::one()
- F::recip()

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -27,22 +27,27 @@ impl<'a, F: Float> DelayLine<'a, F> {
   }
 }
 
+/// Simple delay effect with 3 parameters: the delay amount, the amount of feedback and dry/run mix.
```
```suggestion
/// Simple delay effect with 3 parameters: the delay amount, the amount of feedback and dry/wet mix.
```

chris-zen committed:
fix typo in comment

chris-zen commented on kiro-synth-core/src/effects/delay.rs:
```diff
@@ -0,0 +1,131 @@
+  feedback: F,
+  /// The dry/wet proportion. Values from 0.0 (dry) to 1.0 (wet)
+  mix: F,
+  
```
```suggestion
```

chris-zen committed:
fix fmt
